src/main/java/org/openrewrite/staticanalysis/EqualsAvoidsNullVisitor.java

```diff
@@ -17,84 +17,123 @@
+    }
 
-            if (m.getArguments().get(0).getType() == JavaType.Primitive.Null) {
-                return new J.Binary(Tree.randomId(), m.getPrefix(), Markers.EMPTY,
-                        m.getSelect(),
-                        JLeftPadded.build(J.Binary.Type.Equal).withBefore(Space.SINGLE_SPACE),
-                        m.getArguments().get(0).withPrefix(Space.SINGLE_SPACE),
-                        JavaType.Primitive.Boolean);
-            } else {
-                m = m.withSelect(((J.Literal) m.getArguments().get(0)).withPrefix(m.getSelect().getPrefix()))
-                        .withArguments(singletonList(m.getSelect().withPrefix(Space.EMPTY)));
+    private static J.Binary literalsFirstInComparisonsNull(J.MethodInvocation m, Expression firstArgument) {
+        return new J.Binary(Tree.randomId(),
+                m.getPrefix(),
+                Markers.EMPTY,
+                requireNonNull(m.getSelect()),
+                build(J.Binary.Type.Equal).withBefore(SINGLE_SPACE),
+                firstArgument.withPrefix(SINGLE_SPACE),
+                JavaType.Primitive.Boolean);
+    }
+
+    private static J.MethodInvocation literalsFirstInComparisons(J.MethodInvocation m, Expression firstArgument) {
+        return m.withSelect(firstArgument.withPrefix(requireNonNull(m.getSelect()).getPrefix()))
+                .withArguments(singletonList(m.getSelect().withPrefix(Space.EMPTY)));
+    }
+
+    private void handleBinaryExpression(J.MethodInvocation m, J.Binary binary) {
+        if (binary.getOperator() == J.Binary.Type.And &&
+            binary.getLeft() instanceof J.Binary) {
+            J.Binary left = (J.Binary) binary.getLeft();
+            if (isNullLiteral(left.getLeft()) &&
+                matchesSelect(left.getRight(), requireNonNull(m.getSelect())) ||
+                isNullLiteral(left.getRight()) &&
+                matchesSelect(left.getLeft(), requireNonNull(m.getSelect()))) {
+                doAfterVisit(new RemoveUnnecessaryNullCheck<>(binary));
             }
         }
-
-        return m;
     }
 
     private boolean isNullLiteral(Expression expression) {
         return expression instanceof J.Literal && ((J.Literal) expression).getType() == JavaType.Primitive.Null;
     }
 
     private boolean matchesSelect(Expression expression, Expression select) {
-        return expression.printTrimmed(getCursor()).replaceAll("\\s", "").equals(select.printTrimmed(getCursor()).replaceAll("\\s", ""));
+        return expression.printTrimmed(getCursor()).replaceAll("\\s", "")
+                .equals(select.printTrimmed(getCursor()).replaceAll("\\s", ""));
     }
 
     private static class RemoveUnnecessaryNullCheck<P> extends JavaVisitor<P> {
+
         private final J.Binary scope;
-        boolean done;
 
-        @Override
-        public @Nullable J visit(@Nullable Tree tree, P p) {
-            if (done) {
-                return (J) tree;
-            }
-            return super.visit(tree, p);
-        }
```

Comment:
This visit method seems to have been deliberately added in https://github.com/openrewrite/rewrite-static-analysis/commit/a330fe97f5f04019d8cc5c87d77c351a6e88ca3d ; what made you remove it from here now?

Reply:
Was a mistake, thanks for the finding.

Actually i would be nice if this is tested so i would have known earlier.